common/scala/src/main/scala/whisk/common/tracing/TraceUtil.scala

```diff
@@ -0,0 +1,130 @@
+
+    var requestCounter: AtomicInteger = new AtomicInteger(1);
+
+    def init (actorSystem: ActorSystem): TracingExtensionImpl = {
+
+      if(traceVar == null)
+        traceVar = TracingExtension.apply(actorSystem);
+
+      return trace;
+    }
+
+  /**
+    * Start a Trace for given service.
+    * @param serviceName Name of Service to be traced
+    * @param transactionId transactionId to which this Trace belongs.
+    * @return TracedRequest which provides details about current service being traced.
+    */
+  def startTrace(serviceName: String, transactionId: TransactionId): Unit = {
+    var tracedRequest: TracedRequest = traceMap.get(transactionId.meta.id)
+    if(tracedRequest == null){
+      var tracedRequest = createTracedRequest(serviceName)
+      var metadata = trace.sample(tracedRequest.getRequest(), serviceName)
+      if(metadata != None){
+        trace.start(tracedRequest.getRequest(), serviceName)
+        tracedRequest = new TracedRequest(tracedRequest.getRequest(), null, metadata)
+        traceMap.put(transactionId.meta.id, tracedRequest)
+      }
+    }
+    else
+      startChildTrace(serviceName, tracedRequest, transactionId)
+  }
+
+  /**
+    * Start a Child Trace
+    * @param serviceName Name of service to be traced
+    * @param parent Parent TracedRequest to continue span propagation and create span hierarchy.
+    * @param transactionId transactionId to which this Trace belongs.
+    * @return TracedRequest which provides details about current service being traced.
+    */
+  def startChildTrace(serviceName: String, parent: TracedRequest, transactionId: TransactionId): Unit = {
+    val request: TracedRequest = createTracedRequest(serviceName)
+    var metadata =  trace.sample(request.getRequest(), serviceName)
+    if(metadata != None){
+      trace.createChild(request.getRequest(), parent.getRequest())
+      trace.start(request.getRequest(), serviceName)
+      val tracedRequest = new TracedRequest(request.getRequest(), parent, metadata)
+      traceMap.put(transactionId.meta.id, tracedRequest)
+    }
+
+  }
+
+  /**
+    * Finish a Trace associated with given transactionId.
+    * @param transactionId
+    */
+  def finish(transactionId: TransactionId): Unit = {
+    val request: TracedRequest = traceMap.get(transactionId.meta.id)
+    if(request != null){
+      trace.record(request.getRequest(), TracingAnnotations.ServerSend)
+      if(request.getParent() == null)
+        traceMap.remove(transactionId.meta.id)
+      else
+        traceMap.put(transactionId.meta.id, request.getParent())
+    }
+  }
+
+  def error(transactionId: TransactionId, t: Throwable) : Unit = {
+    val request: TracedRequest = traceMap.get(transactionId.meta.id)
+    if(request != null){
+      trace.record(request.getRequest(), t)
+    }
+  }
+
+  private def createTracedRequest(serviceName: String, metadata: Option[SpanMetadata] = None) : TracedRequest = {
+    val headers: Map[String, Integer] = new HashMap[String, Integer]
```

Comment:
I would use normal B3 for trace propagation default even if using a custom envelope. This will be a lot more sensible if someone is inspecting messages. The actual size of encoding this way isn't likely a concern, and being able to re-use docs and understanding would be worth it imho

https://github.com/openzipkin/b3-propagation